### PR TITLE
MAYA-114241 - Manage the edit-ability of an attribute from Maya's att…

### PR DESCRIPTION
**Pull request by seando-adsk**
#### MAYA-114241 - Manage the edit-ability of an attribute from Maya's attribute editor

* Add lock/unlock attribute menu and status to array attributes.

### CODE REVIEW

**Comment by seando-adsk** [comment]
This will add the lock/unlock attribute menu item to the array attributes (custom widget in our AE template).
image

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 1]
```diff
@@ -209,7 +211,33 @@ def _onInstanceableChanged(self, value):
 # Custom control for all array attribute.
 class ArrayCustomControl(object):
 
-    def __init__(self, prim, attrName, useNiceName):
+    class ArrayAEPopup(attributes.AEPopupMenu):
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:214]: Override this class from Maya so we can inject the copy/print menu items.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 2]
```diff
@@ -209,7 +211,33 @@ def _onInstanceableChanged(self, value):
...
+        def _printToScriptEditor(self):
+            MGlobal.displayInfo(str(self.values))
+
+        COPY_ACTION  = ('Copy Attribute Value', _copyAttributeValue, [])
+        PRINT_ACTION = ('Print to Script Editor', _printToScriptEditor, [])
+
+        HAS_VALUE_MENU = [COPY_ACTION, PRINT_ACTION]
+
+        def _buildMenu(self, addItemCmd):
+            super(ArrayCustomControl.ArrayAEPopup, self)._buildMenu(addItemCmd)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:234]: This adds the base menu items from Maya, which is the lock/unlock attribute.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 3]
```diff
@@ -209,7 +211,33 @@ def _onInstanceableChanged(self, value):
...
+            if self.hasValue:
+                cmds.menuItem(divider=True, parent=self.popupMenu)
+                self._buildFromActions(self.HAS_VALUE_MENU, addItemCmd)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:237]: Then if this array attribute has a value, add the other menu items.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 4]
```diff
@@ -236,16 +264,29 @@ def onCreate(self, *args):
...
 
     def onReplace(self, *args):
         pass
 
+    def updateUi(self, attr, uiControlName):
+        with ParentManager(uiControlName):
+            bgClr = attributes.getAttributeColorRGB(self.ufeAttr)
+            if bgClr:
+                isLocked = attributes.isAttributeLocked(self.ufeAttr)
+                cmds.textField(attrTypeFld, edit=True, backgroundColor=bgClr)
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:282]: Called when attribute changes - so we can update the locked status.

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py (hunk 1 above)
Comment (seando-adsk) [reply lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:214]: I realized the AEPopupMenu is only available in latest Maya, so I'll need some logic here to use new vs old way.

**Review by ppt-adsk — APPROVED** [review]

**Review by ppt-adsk — APPROVED** [review]

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 5]
```diff
@@ -21,7 +21,15 @@
...
 import mayaUsd.lib as mayaUsdLib
 import maya.internal.common.ufe_ae.template as ufeAeTemplate
-from maya.common.ui import LayoutManager
+
+try:
+    # This helper class was only added recently to Maya.
+    import maya.internal.ufeSupport.attributes as attributes
+    hasAEPopupMenu = 'AEPopupMenu' in dir(attributes)
+except:
+    hasAEPopupMenu = False
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:30]: Set a variable here which says whether or not we have the new AEPopupMenu class (from Maya).

**Review by ppt-adsk — APPROVED** [review]

File: lib/mayaUsd/resources/ae/usdschemabase/ae_template.py [hunk 6]
```diff
@@ -209,7 +217,34 @@ def _onInstanceableChanged(self, value):
 # Custom control for all array attribute.
 class ArrayCustomControl(object):
 
-    def __init__(self, prim, attrName, useNiceName):
+    if hasAEPopupMenu:
```
Comment (seando-adsk) [inline lib/mayaUsd/resources/ae/usdschemabase/ae_template.py:220]: Only create this class if we have the base class

**Review by ppt-adsk — APPROVED** [review]

**Pull request merged**